### PR TITLE
Add gradient accumulation support for all backends, and enable optimizer EMA for JAX and torch

### PR DESCRIPTION
**Pull request by fchollet**

### CODE REVIEW

**Comment by fchollet** [comment]
@qlzh727 please review this PR -- in particular, check whether the EMA variables and the gradient accumulators are updated in a way that is correct in a tf.distribute context

**Comment by fchollet** [comment]
@haifeng-jin can you please review the changes in `Nadam` and provide your analysis on why the torch tests are failing? I think it has to do with the fact that torch has its own custom Nadam update step.

**Comment by haifeng-jin** [comment]
Nice!

File: keras/backend/tensorflow/optimizer.py [hunk 1]
```diff
@@ -97,25 +97,21 @@ def weight_decay_fn(variable):
...
             self._distribution_strategy,
-            grads_and_vars,
+            list(zip(grads, trainable_variables)),
+            learning_rate,
         )
 
-    def _distributed_apply_gradients_fn(
-        self, distribution, grads_and_vars, **kwargs
+    def _distributed_tf_update_step(
+        self, distribution, grads_and_vars, learning_rate
```
Comment (qlzh727) [inline keras/backend/tensorflow/optimizer.py:113]: Seems that the learning_rate is not used here, the value in the apply_grad_to_updatew_var was retrieved from self._get_current_learning_rate(). Did I miss anything?

File: keras/backend/jax/optimizer.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+            for n_g_acc, g_acc in zip(new_g_accs, self._accumulated_gradients):
+                g_acc.assign(n_g_acc)
+
+        else:
+            self._backend_update_step(
+                grads, trainable_variables, self.learning_rate
+            )
+
+        if self.use_ema:
```
Comment (qlzh727) [inline keras/backend/jax/optimizer.py:77]: Just curious, does the existing jax optimizer support ema? or its from the base_optimizer?

File: keras/optimizers/optimizer_test.py [hunk 3]
```diff
@@ -7,6 +7,36 @@
 
 
 class OptimizerTest(testing.TestCase):
```
Comment (qlzh727) [inline keras/optimizers/optimizer_test.py:9]: Can we also update the test in tensorflow/optimizer_distribute_test.py for distribution related test case? (for ema and gradient accumulation)

File: keras/backend/jax/optimizer.py (hunk 2 above)
Comment (fchollet) [reply keras/backend/jax/optimizer.py:77]: Now JAX supports the feature -- the new unit tests check it. The previous JAX optimizer (before this PR) didn't -- only TF did.

File: keras/optimizers/base_optimizer.py [hunk 4]
```diff
@@ -304,19 +327,100 @@ def apply(self, grads, trainable_variables=None):
...
+        for g_acc in self._accumulated_gradients:
+            g_acc.assign(np.zeros(g_acc.shape, dtype=g_acc.dtype))
+
+    def _backend_increment_gradient_accumulators(self, grads):
+        new_g_accs = [
+            (grads[i] + self._accumulated_gradients[i])
+            for i in range(len(grads))
+        ]
+        for n_g_acc, g_acc in zip(new_g_accs, self._accumulated_gradients):
+            g_acc.assign(n_g_acc)
```
Comment (qlzh727) [inline keras/optimizers/base_optimizer.py:423]: In the tf.distribute context, I think this probably should use https://www.tensorflow.org/api_docs/python/tf/distribute/StrategyExtended#update

Comment (fchollet) [reply keras/optimizers/base_optimizer.py:423]: Ok, but what about `self.iterations`? We update with a simple `assign`. When is it ok to use assign and when is it not?

File: keras/backend/tensorflow/optimizer_distribute_test.py [hunk 5]
```diff
@@ -163,30 +163,24 @@ def test_ema(self):
     def test_gradient_accumulation(self):
         with self.strategy.scope():
             v = backend.Variable([[1.0, 2.0], [3.0, 4.0]])
-            grads = backend.convert_to_tensor([[1.0, 1.0], [1.0, 1.0]])
+            grads = backend.convert_to_tensor([[1.0, 1.0], [2.0, 2.0]])
```
Comment (qlzh727) [inline keras/backend/tensorflow/optimizer_distribute_test.py:166]: Thanks for adding the test, can we test with the grad that has different value on each replica? 

You can create a distribute value via https://www.tensorflow.org/api_docs/python/tf/distribute/DistributedValues

Comment (fchollet) [reply keras/backend/tensorflow/optimizer_distribute_test.py:166]: Best I can tell, it is not possible to call an optimizer (or any other Keras function) with `DistributeValues`. The reason is that `DistributeValues` is a stand-in for a tensor, but it does not implement the tensor API (no `.shape`, no `.dtype`, etc).

**Review by qlzh727 — APPROVED** [review]
